src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala

```diff
@@ -201,58 +202,82 @@ object PrecompiledContracts {
           safeAdd(safeAdd(totalLengthBytes, baseLength), expLength)
         )
 
-      val multComplexity = getMultComplexity(math.max(baseLength, modLength))
+      if (ethFork >= EthForks.Berlin || etcFork >= EtcForks.Magneto)
+        PostEIP2565Cost.calculate(baseLength, modLength, expLength, expBytes)
+      else
+        PostEIP198Cost.calculate(baseLength, modLength, expLength, expBytes)
+    }
 
-      val adjExpLen = adjExpLength(expBytes, expLength)
+    object PostEIP198Cost {
+      private val GQUADDIVISOR = 20
 
-      multComplexity * math.max(adjExpLen, 1) / GQUADDIVISOR
-    }
+      def calculate(baseLength: Int, modLength: Int, expLength: Int, expBytes: ByteString): BigInt = {
+        val multComplexity = getMultComplexity(math.max(baseLength, modLength))
+        val adjusted = adjustExpLength(expBytes, expLength)
+        multComplexity * math.max(adjusted, 1) / GQUADDIVISOR
+      }
 
-    private def adjExpLength(expBytes: ByteString, expLength: Int): Long = {
-      val expHead =
-        if (expLength <= lengthBytes)
-          expBytes.padToByteString(expLength, 0.toByte)
+      private def getMultComplexity(x: BigInt): BigInt = {
```

Comment:
maybe a comment explaining all those numbers? I wouldn't know how to change that piece because I don't know where all those numbers are coming from

Reply:
hence the name of the object (PostEIP198Cost). I can quote the EIP but I think this is too complex to just explain away in a comment - it needs to be read at the source. I'll add a link to the EIPs though.